aggregation/src/lib.rs

```diff
@@ -52,45 +53,309 @@ alloy_sol_types::sol! {
+    /// (i.e. a leaf) has a max_depth value of 0.
+    max_depth: u8,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum Error {
+    #[error("Merkle mountain range is finalized")]
+    FinalizedError,
+    #[error("Merkle mountain range is empty")]
+    EmptyError,
+    #[error("decoding error: {0}")]
+    DecodingError(#[from] DecodingError),
 }
 
-impl GuestInput {
-    #[inline]
-    pub fn image_id(&self) -> Digest {
-        match self {
-            GuestInput::Singleton { self_image_id, .. } => *self_image_id,
-            GuestInput::Join { self_image_id, .. } => *self_image_id,
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum DecodingError {
+    #[error("invalid bitmap")]
+    InvalidBitmap,
+    #[error("unexpected end of byte stream")]
+    UnexpectedEnd,
+    #[error("trailing bytes")]
+    TrailingBytes,
+}
+
+impl MerkleMountainRange {
+    pub fn empty() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Construct a new [MerkleMountainRange] in a finalized state, given a root.
+    pub fn new_finalized(root: Digest) -> Self {
+        Self(vec![Peak {
+            max_depth: u8::MAX,
+            digest: root,
+        }])
+    }
+
+    /// Push a new leaf onto the Merkle mountain range.
+    pub fn push(&mut self, leaf: Digest) -> Result<(), Error> {
+        self.push_peak(Peak {
+            digest: leaf,
+            max_depth: 0,
+        })
+    }
+
+    fn push_peak(&mut self, new_peak: Peak) -> Result<(), Error> {
+        // If the peak has a max-depth of 255, then the mountain range is finalized and no new
+        // peaks can be pushed to it. Note that this state can only be achieved by calling
+        // `finalize` since it is computationally infeasible to push 2^256 nodes onto the MMR,
+        // although it is theoretically consistent that if an MMR reached a state of being a
+        // single peak with max depth value of 255, it would be naturally finalized.
+        if self.is_finalized() {
+            return Err(Error::FinalizedError);
+        }
+        match self.0.last() {
+            // If the MerkleMountainRange is empty, push the new peak.
+            None => self.0.push(new_peak),
+            // If the tail subtree is larger, push the new subtree onto the end.
+            Some(peak) if peak.max_depth > new_peak.max_depth => {
+                self.0.push(new_peak);
+            }
+            // If the tail subtree is the same size, combine them and recurse.
+            Some(peak) if peak.max_depth == new_peak.max_depth => {
+                // Will never panic, since we've already checked that there is at least one peak.
+                let peak = self.0.pop().unwrap();
+                self.push_peak(Peak {
+                    digest: commutative_keccak256(&peak.digest, &new_peak.digest),
+                    max_depth: peak.max_depth + 1, //TODO this is fallible
```

Comment:
Even though 2^254 is very, very large, we could just add an assert(peak.max_depth < 0xfe) here for clarity.

Reply:
Used `checked_add` here. Thanks for catching this, I meant to address it.
928fccc